datafusion/optimizer/src/analyzer/subquery.rs

```diff
@@ -0,0 +1,340 @@
+                    predicate
+                )));
+            }
+            check_inner_plan(input, is_scalar, is_aggregate, can_contain_outer_ref)
+        }
+        LogicalPlan::Window(window) => {
+            check_mixed_out_refer_in_window(window)?;
+            inner_plan.apply_children(&mut |plan| {
+                check_inner_plan(plan, is_scalar, is_aggregate, can_contain_outer_ref)?;
+                Ok(VisitRecursion::Continue)
+            })?;
+            Ok(())
+        }
+        LogicalPlan::Projection(_)
+        | LogicalPlan::Distinct(_)
+        | LogicalPlan::Sort(_)
+        | LogicalPlan::CrossJoin(_)
+        | LogicalPlan::Union(_)
+        | LogicalPlan::TableScan(_)
+        | LogicalPlan::EmptyRelation(_)
+        | LogicalPlan::Limit(_)
+        | LogicalPlan::Values(_)
+        | LogicalPlan::Subquery(_)
+        | LogicalPlan::SubqueryAlias(_) => {
+            inner_plan.apply_children(&mut |plan| {
+                check_inner_plan(plan, is_scalar, is_aggregate, can_contain_outer_ref)?;
+                Ok(VisitRecursion::Continue)
+            })?;
+            Ok(())
+        }
+        LogicalPlan::Join(Join {
+            left,
+            right,
+            join_type,
+            ..
+        }) => match join_type {
+            JoinType::Inner => {
+                inner_plan.apply_children(&mut |plan| {
+                    check_inner_plan(
+                        plan,
+                        is_scalar,
+                        is_aggregate,
+                        can_contain_outer_ref,
+                    )?;
+                    Ok(VisitRecursion::Continue)
+                })?;
+                Ok(())
+            }
+            JoinType::Left | JoinType::LeftSemi | JoinType::LeftAnti => {
+                check_inner_plan(left, is_scalar, is_aggregate, can_contain_outer_ref)?;
+                check_inner_plan(right, is_scalar, is_aggregate, false)
+            }
+            JoinType::Right | JoinType::RightSemi | JoinType::RightAnti => {
+                check_inner_plan(left, is_scalar, is_aggregate, false)?;
+                check_inner_plan(right, is_scalar, is_aggregate, can_contain_outer_ref)
+            }
+            JoinType::Full => {
+                inner_plan.apply_children(&mut |plan| {
+                    check_inner_plan(plan, is_scalar, is_aggregate, false)?;
+                    Ok(VisitRecursion::Continue)
+                })?;
+                Ok(())
+            }
+        },
+        _ => Err(DataFusionError::Plan(
+            "Unsupported operator in the subquery plan.".to_string(),
+        )),
+    }
+}
+
+fn contains_outer_reference(inner_plan: &LogicalPlan) -> bool {
+    inner_plan
+        .expressions()
+        .iter()
+        .any(|expr| expr.contains_outer())
+}
+
+fn check_aggregation_in_scalar_subquery(
+    inner_plan: &LogicalPlan,
+    agg: &Aggregate,
+) -> Result<()> {
+    if agg.aggr_expr.is_empty() {
```

Comment:
it seems like you might be able to include this logic in `LogicalPlan::max_rows()` and avoid having a special check for aggregates

Reply:
Yes.